Use SQLAlchemy core for database access and query/dialect

kinow opened the pull request:
These changes close #3360

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included. (WIP: will add one if it works)
- [ ] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/XXXX. (WIP: it may need new/updated documentation)

oliver-sanders commented:
It looks like there might be scope to document the database (e.g. via the `doc` field of `Column` https://docs.sqlalchemy.org/en/13/core/metadata.html). There may even be some Sphinx infrastructure to support this https://pypi.org/project/sphinxcontrib-sqlalchemyviz/.

hjoliver commented:
@dpmatthews flagged this as maybe an unnecessary/distracting change given current priorities and stretched resources. However, he may not have realized @matthewrmshin was on board with it (and @kinow said low priority and only pinned it to Cylc 9 anyway).

However, we seemed to have a consensus on the team that this is an improvement **and** it makes supporting other DBs easier in future.  So I think **if** it is **easy** to finish this off so that restarts work, and the DB content doesn't change at all (so downstream tools won't be affected at all), we should merge it so we don't have to redo it next year.  (And then try to avoid other "distracting" changes until Cylc 8 is out :-)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

kinow commented:
Unit tests passing locally, now will let Travis go through functional tests, and check what else is broken.

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

kinow commented on cylc/flow/suite_db_mgr.py:
```diff
@@ -58,18 +61,18 @@ class SuiteDatabaseManager(object):
+    TABLE_TASK_OUTPUTS: Table = task_outputs
+    TABLE_TASK_STATES: Table = task_states
+    TABLE_TASK_TIMEOUT_TIMERS = task_timeout_timers
+    TABLE_XTRIGGERS = xtriggers
```
Fixed

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

kinow commented:
Will spend time on this PR only on spare time/rainy holidays now :grimacing: as this is not a priority for the next release I think.

But during this summer break I managed to fix several functional tests. Some of the errors I had to fix in this PR:

- Cylc 7/8 at the moment has at least one SQL statement that is not easy to port to other databases. When you run the workflow, it creates `task_action_timers`. Then in the next run of the main loop, it will create more `task_action_timers`, but some conflict with PK's in the DB. So what the code is doing right now is `INSERT OR REPLACE`. Other DB's have the `INSERT ... ON CONFLICT` (Postgres), `INSERT + EXCEPTION` or `INSERT ... WHERE NOT EXISTS` (Oracle). Not sure if it is supported by SqlAlchemy
- I translated some queries that had `LIMIT 1` to `fetchall` (equivalent to `SELECT` without the `LIMIT 1`), then later had to replace by `fetchone`
- SqlAlchemy by default returns tuples with data only. Parts of our code need the data and the column name in a dict (e.g. select_task_jobs)

One by-product of this change that I liked, is how you can toggle SqlAlchemy logging, to compare the queries executed. It does pretty-printing too, and prints the arguments used in the next line. Example of the logs in a transaction.

```
2020-01-03 11:21:19,281 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-01-03 11:21:19,281 INFO sqlalchemy.engine.base.Engine INSERT INTO inheritance (namespace, inheritance) VALUES (?, ?)
2020-01-03 11:21:19,282 INFO sqlalchemy.engine.base.Engine ('root', '["root"]')
2020-01-03 11:21:19,282 INFO sqlalchemy.engine.base.Engine INSERT INTO inheritance (namespace, inheritance) VALUES (?, ?)
2020-01-03 11:21:19,282 INFO sqlalchemy.engine.base.Engine ('prep', '["prep", "root"]')
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine INSERT INTO inheritance (namespace, inheritance) VALUES (?, ?)
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine ('foo', '["foo", "root"]')
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine INSERT INTO inheritance (namespace, inheritance) VALUES (?, ?)
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine ('bar', '["bar", "root"]')
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine DELETE FROM suite_params
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine ()
2020-01-03 11:21:19,283 INFO sqlalchemy.engine.base.Engine INSERT INTO suite_params ("key", value) VALUES (?, ?)
2020-01-03 11:21:19,284 INFO sqlalchemy.engine.base.Engine ('uuid_str', '41e3c9dc-8e29-48d1-ab9b-2305122d149c')
2020-01-03 11:21:19,284 INFO sqlalchemy.engine.base.Engine INSERT INTO suite_params ("key", value) VALUES (?, ?)
...
...
2020-01-03 11:21:19,289 INFO sqlalchemy.engine.base.Engine INSERT INTO task_states (name, cycle, time_created, time_updated, submit_num, status) VALUES (?, ?, ?, ?, ?, ?)
2020-01-03 11:21:19,289 INFO sqlalchemy.engine.base.Engine ('bar', '20130808T0000Z', '2020-01-02T22:21:19Z', '2020-01-02T22:21:19Z', 0, 'waiting')
2020-01-03 11:21:19,290 INFO sqlalchemy.engine.base.Engine UPDATE task_states SET time_updated=?, submit_num=?, status=? WHERE task_states.cycle = ? AND task_states.name = ?
2020-01-03 11:21:19,290 INFO sqlalchemy.engine.base.Engine ('2020-01-02T22:21:19Z', 1, 'ready', '20130808T0000Z', 'prep')
2020-01-03 11:21:19,290 INFO sqlalchemy.engine.base.Engine DELETE FROM task_timeout_timers
2020-01-03 11:21:19,291 INFO sqlalchemy.engine.base.Engine ()
2020-01-03 11:21:19,291 INFO sqlalchemy.engine.base.Engine DELETE FROM xtriggers
2020-01-03 11:21:19,291 INFO sqlalchemy.engine.base.Engine ()
2020-01-03 11:21:19,291 INFO sqlalchemy.engine.base.Engine COMMIT
```

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/dbstatecheck.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/dbstatecheck.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/dbstatecheck.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/dbstatecheck.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

hjoliver commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/rundb.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/dbstatecheck.py  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4151565&bid=14238323)

kinow commented:
Leaving the PR open would cause noise and eventually run into more conflicts. I think it's easier to close it now.

I've added the requirements in the issue that it was closing. Also added a link to [this gist](https://gist.github.com/kinow/9013c8ad1d13d4f50d6e1df769bf49c7) where I explain how I ran `five` with Postgres :tada: 

In the next days will update that text and write a blog post so I remember everything I have in mind. Also learned a bit how the database works in Cylc, the role of suite_db_mgr, learned that dbtstatecheck also interfaces with the database, that the DAOs try to stay in sync, that suite_db_mgr and DAOs both keep a list of operations to be executed, which the WFS gives it a chance every loop to batch-run all the statements. Will write about this too later :nerd_face:
